### PR TITLE
Guard Cells with PSATD: Fill With Inverse FFT, Damp Fields

### PR DESCRIPTION
**Pull request by EZoni**
This PR adds two features for PSATD simulations:

- fill the guard cells when we compute inverse FFTs;
- extend the field damping in the guard cells to Cartesian geometry.

### CODE REVIEW

File: Source/WarpX.cpp [hunk 1]
```diff
@@ -880,13 +880,12 @@ WarpX::ReadParameters ()
                 "psatd.update_with_rho must be equal to 1 for comoving PSATD");
         }
 
-#   ifdef WARPX_DIM_RZ
-        if (!Geom(0).isPeriodic(1)) {
+        constexpr int zdir = (AMREX_SPACEDIM == 2) ? 1 : 2;
```
Comment (dpgrote) [inline Source/WarpX.cpp]: A minor comment - elsewhere in the code the expression (AMREX_SPACEDIM-1) is used to get the z direction.

Comment (EZoni) [reply Source/WarpX.cpp]: Oh yes, absolutely, you're right. Will change that, have various commits to come, to fix various things. Thank you!

**Comment by EZoni** [comment]
@dpgrote

I think I will have to modify the function `WarpX::DampFieldsInGuards` in order to make it work properly in Cartesian geometry.

The issue is the following. In the current implementation we loop over the box extracted from the MultiFab `Efield[0]` (that is, Ex or Er in RZ). This is not a problem in RZ geometry because all MultiFabs are cell-centered. However, in Cartesian geometry and when `do_nodal = 0` (Yee grid), some fields are nodal in some directions while other fields aren't. This causes out-of-bound access in the arrays, when we try to access the last nodal point that might be missing.

Here's an example of the situation (`tilebox` is the box used to loop over). For instance, we cannot access the array extracted from Ez or Bx or By with index 136 in z (meaning the second index):
```
tilebox = ((-8,-8) (135,136) (0,1))
(*Efield[1])[mfi].box() = ((-8,-8) (136,136) (1,1))
(*Efield[2])[mfi].box() = ((-8,-8) (136,135) (1,0))
(*Bfield[0])[mfi].box() = ((-8,-8) (136,135) (1,0))
(*Bfield[1])[mfi].box() = ((-8,-8) (135,135) (0,0))
(*Bfield[2])[mfi].box() = ((-8,-8) (135,136) (0,1))
```

So, I would like to modify the function by simply passing to it one MultiFab at a time. So that it loops over the boxes extracted from each MultiFab independently. This would not change the behavior of the function in RZ geometry and would fix the issue in Cartesian geometry.

I can do this in a separate PR. Would that be a problem or interfere with the work you were planning to do on this function?

**Comment by dpgrote** [comment]
Hi @EZoni Yes, that is a problem when there is staggering. I would do the triple ParallelFor over the three boxes, i.e. `ParallelFor(tbx, tby, tbz, `, doing one for the Es and one for the Bs. It will be changing the same code as I did but it will be straightforward to fix the conflicts so don't worry about that.

A little care might be needed to determine the value of the damping factor for differently centered fields, whether to include the 0.5*dz in zcell for the staggered fields. I don't know if it matters.

**Comment by EZoni** [comment]
@dpgrote OK, thank you for the suggestion. I'll use the triple `ParallelFor` then. And regarding the damping factor, I think I'll keep it like this in the separate preparatory PR and then see if I have to fix it once I enable the damping for Cartesian geometry in this PR here.

**Comment by EZoni** [comment]
I tested this PR on two 3D test cases, LWFA ("high" gamma boost around 30) and PWFA ("low" gamma boost around 4) and it seems to me to work as expected (see plots below). Both test cases ran on 24 GPUs on Summit, with domain decomposition both in (x,y) and along z.

Hence, I think the PR is ready for a first round of review. In the meantime I will keep testing it more.

3D LWFA test case, Ex field,
(16,16,16) guard cells in the nodal case,
(8,8,16) guard cells in the hybrid case:
<p align="center">
    
    
</p>

3D PWFA test case, Ez field with beam,
(26,26,16) guard cells in the nodal case,
(8,8,16) guard cells in the hybrid case:
<p align="center">
    
    
</p>

**Review by dpgrote — COMMENTED** [review]
This looks good, thanks! My only suggestion is to use more descriptive names for the two new routines set_tilebox and damp_fields. Perhaps `constrain_tilebox_to_guards` and 'damp_fields_in_guards`.

**Review by dpgrote — APPROVED** [review]
Thanks for the update!

**Comment by EZoni** [comment]
@dpgrote Thank you for the review and especially for the initial suggestions on how to modify and extend the damping functions properly.

**Comment by RemiLehe** [comment]
Thanks a lot for this PR!

**Pull request merged**